Add git bash navigation tips page (#24)
dfe-analytical-services/git-academy: git-bash-hints.qmd

---
title: "Git Bash Hints"
editor: visual
---

This page will cover a few useful commands that you can use in Git Bash. 

:::{.callout-note}

You will see many generalised paths, directory names, file names, etc. on this page. For example, you will see something like `cd /path_to/working_directory/you_want` where you would need to replace the `/path_to/working_directory/you_want` part with a path that is applicable to your work.

:::

## Checking and Navigating Directories

The commands in this section allow you to check and change your current working directory, moving between different folders in your file system.

`pwd`: This stands for "print working directory", so it will show you your current working directory for Git.

You can use it by entering the following in the terminal: 

> ```{r, eval = FALSE}

Reviewer: These should be switched to bash rather than r, although I'm not sure it will add any extra rendering off the back of that. Also, I don't think you need the comma.

```suggestion
```{bash eval = FALSE}
```

pwd

```

`cd`: This stands for "change directory", so it will point to where you want your directory to be.

> To change your working directory, we would amend the code below to reflect the path of the working directory we want and enter it into the terminal: 

Reviewer: Worth noting you can use relative or absolute file paths here.

> ```{r, eval = FALSE}

Reviewer: Another on to switch to bash eval=FALSE

cd /path_to/working_directory/you_want

```

 
## Viewing Files and Directories

The commands in this section help you list and view the contents of directories, providing detailed information about files and sub-directories.

`ls`: This stands for "listing".

To list all files or sub-directories, we would type the following in the terminal: 

> ```{r, eval = FALSE}

Reviewer: bash

ls

```

To list all files or sub-directories, including hidden files, we would type the following in the terminal: 

> ```{r, eval=FALSE}

Reviewer: bash

ls -a

```

> 

Reviewer: Another one I like is `ls -ltrh`, which lists all files with dates amd human readable sizes in reverse-date order.

## Creating and Removing Files and Directories

### Creating directories 

The commands in this section enable you to create new directories, remove files, and delete empty directories.

`mkdir`: This stands for "make directory".

To create a directory, we would amend the code below to reflect the path we want and the new directory name and enter it into the terminal: 

> ```{r, eval=FALSE}

Reviewer: bash

> mkdir /path_to_the_directory_you_want_to_create/new_directory_name

Reviewer: Again worth noting this can be absolute or relative paths

Reviewer: In fact I'd simplify the example to just be realtive, so `mkdir new_directory_name` and say that will create a new directory in your current working directory.

```

### Removing directories 

:::{.callout-important}

Make sure you're in the correct directory/using the correct path before using commands in this section.
:::

`rm`: This stands for "remove" and we can use it to remove files and directories.

To remove a directory and its contents, we would amend the code below to reflect the path and the directory we want and enter it into the terminal: 

> ```{r, eval=FALSE}

Reviewer: bash

> rm -r /path_to_the_directory/directory_name

Reviewer: I'd be tempted to just switch to relative path here. I think that's the more natural way to use these commands. Can probably just remove the first `/`. Also worth adding a word of caution about recursive mode.

```

### Removing files

 
Assuming we're already in the directory we want, we can remove a file from it by amending the code below to reflect the file name we want to remove, and its extension and then enter it into the terminal: 

> ```{r, eval=FALSE}

Reviewer: bash

> rm -r file_name.extention

Reviewer: Don't need recursive mode here given it's just a single file

```

You need to make sure to include the extension of the file in the command, or it will not work. 

## Viewing File Contents

:::{.callout-important}

Make sure you're in the correct directory/using the correct path before using commands in this section.
:::

The commands in this section allow you to display the contents of files, either in full or in parts, for easy reading and inspection.

To display the first 10 lines of a file, we would amend the code below to reflect the file name we want to view and its extension and then enter it into the terminal:

> ```{r, eval=FALSE}

Reviewer: bash

> head file_name.extention

Reviewer: Useful thing to mention here and in tail is that you can set the number of lines to return using e.g.:
`tail -3 filename`
and
`head -12 filename`

```

To display the bottom 10 lines of a file, we would amend the code below to reflect the file name we want to view and its extension and then enter it into the terminal:

> ```{r, eval=FALSE}

Reviewer: bash

tail file_name.extention

```

To display the contents of a file, we would amend the code below to reflect the file name we want to view and its extension and then enter it into the terminal:

> ```{r, eval=FALSE}

Reviewer: bash

cat file_name.extention

```

## Searching and Finding Files

:::{.callout-important}

Make sure you're in the correct directory/using the correct path before using commands in this section.
:::

You can look for files in your repository that match a certain pattern by using `git grep`. We would amend the code below by replacing `"pattern"` with the term or pattern we want to search for in files. 

> ```{r, eval=FALSE}

Reviewer: bash

git grep "pattern"

```

You can utilise all of regex's capabilities to locate precisely what you're looking for because regular expressions are supported.

## Creating Files

:::{.callout-important}

Make sure you're in the correct directory/using the correct path before using commands in this section.
:::

The commands in this section enable you to  create an empty file or update the timestamp of an existing file. We do this by entering the following this command template `touch <file_name.extention>` where you replace file_name with the name of the file you want to create and `extention` with the extension for the type of file you want to create. 

For example, if we want to create a `txt` type file called `example`, then we would type the following in the terminal:

> ```{r, eval=FALSE}

Reviewer: bash

---
title: "BASH Academy"
editor: visual
engine: knitr
execute:
  eval: false
---

This page will cover a few useful commands that you can use in Git BASH.

You will see many generalised paths, directory names, file names, etc. on this page. For example, you will see something like `cd /path_to/working_directory/you_want` where you would need to replace the `/path_to/working_directory/you_want` part with a path that is applicable to your work.

::: callout-important

## OneDrive

> When using Git commands like `find`, `ls` or `grep` in OneDrive, be cautious as OneDrive's file synchronization can interfere with Git operations, potentially leading to conflicts or data corruption. Additionally, OneDrive will download all files to search them, which can consume significant bandwidth and storage. It's advisable to pause OneDrive syncing or use a local directory for Git repositories to avoid these issues.

Reviewer: Just switching this to focus on Bash rather than bringing Git into it as well.
```suggestion
Avoid using Bash commands like `find`, `ls` or `grep` within OneDrive folders unless OneDrive syncing is switched off. OneDrive will download all files to search them, which can consume significant bandwidth and storage. It's advisable to pause OneDrive syncing if working with Bash commands within your local OneDrive folders.
```

:::

## Checking and Navigating Directories

The commands in this section allow you to check and change your current working directory, moving between different folders in your file system.

### Checking directories

> `pwd`: This stands for "print working directory", so it will show you your current working directory for Git.

Reviewer: ```suggestion
`pwd`: This stands for "print working directory", so it will show you your current working directory within the terminal.
```

You can use it by entering the following in the terminal:

```{bash }

> 

Reviewer: Extra line could do with being deleted here
```suggestion
```

```

#### Using a relative path:

Relative paths specify the path in relation to your current directory, allowing you to navigate efficiently within the file system.

##### Navigating into a folder:

```{bash }
cd subfolder

```

-   This command navigates into a folder named `subfolder` inside `my_project`.

-   Resulting path: `/home/user/projects/my_project/subfolder`.

##### Navigating out of a folder (up one level):

```{bash }
cd ..
```

This command moves up one level from my_project to projects. Resulting path: `/home/user/projects`

### Auto complete

Tab completion is a feature that allows you to auto-complete file and directory names by pressing the Tab key. If there are multiple matches, pressing Tab twice will list all possible completions.

```{bash }

# Type 'cd /path/to/dir' and press 'Tab' to auto-complete the directory name.
cd /path/to/dir

```

## Viewing Files and Directories

The commands in this section help you list and view the contents of directories, providing detailed information about files and sub-directories.

`ls`: This stands for "listing".

To list all files or sub-directories, we would type the following in the terminal:

```{bash }

> 

Reviewer: Removing white space
```suggestion
```

ls

> 

Reviewer: Removing whitespace
```suggestion
```

```

To list all files or sub-directories, including hidden files, we would type the following in the terminal:

```{bash }

> 

Reviewer: ```suggestion
```

ls -a

> 

Reviewer: ```suggestion
```

```

You can add more options to `ls` to control how to list files. We can use `-ltrh` options modify the output of `ls`:

-   `l`: Use a long listing format.
-   `t`: Sort by modification time, newest first.
-   `r`: Reverse the order while sorting.
-   `h`: Display sizes in a human-readable format.

To list the contents of the specified directory in long format, sorted by modification time (newest last), in reverse order, with human-readable file sizes, we would type the following in the terminal: 

```{bash }

> 

Reviewer: ```suggestion
```

ls -ltrh /path/to/directory

> 

Reviewer: ```suggestion
```

```

## Creating and Removing Files and Directories

::: callout-important
Make sure you're in the correct directory/using the correct path before using commands in this section.
:::

### Creating directories

The commands in this section enable you to create new directories, remove files, and delete empty directories.

`mkdir`: This stands for "make directory".

To create a directory, we would amend the code below to reflect the new directory name and enter it into the terminal:

```{bash }

> 

Reviewer: ```suggestion
```

mkdir new_directory_name

> 

Reviewer: ```suggestion
```

```

This will create a new directory in your current working directory.

### Creating files

#### `echo`

The echo command is used to display a line of text or a variable value. It can also be used to create files by redirecting the output to a file.

To create a file named hello.txt and writes "Hello, World!" into it, we would enter the following in the terminal:

```{bash }

> 

Reviewer: ```suggestion
```

echo "Hello, World!" > hello.txt

> 

Reviewer: ```suggestion
```

> 

Reviewer: ```suggestion
```

```

To append text to an existing file instead of overwriting it, you can use the \>\> operator:

```{bash }

> 

Reviewer: ```suggestion
```

echo "This is additional text." >> hello.txt

> 

Reviewer: ```suggestion
```

```

This command adds "This is additional text." to the end of the hello.txt file without overwriting its existing content.

> `touch`

Reviewer: Missed heading flag
```suggestion
#### `touch`
```

The command in this section enables you to create an empty file or update the timestamp of an existing file. We do this by entering the following this command template `touch <file_name.extention>` where you replace file_name with the name of the file you want to create and `extention` with the extension for the type of file you want to create.

For example, if we want to create a `txt` type file called `example`, then we would type the following in the terminal:

```{bash }

> 

Reviewer: ```suggestion
```

touch example.txt

> 

Reviewer: ```suggestion
```

For example, if I had a file called `example.txt` that already existed in my directory and its timestamp for when it was last modified was 27/01/2025 at 09:00 AM. Then I used the command `touch example.txt` on 28/01/2024 at 10:30 AM, the timestamp for that file will be modified to reflect the time I entered the command (28/01/2024 at 10:30 AM).

### Removing directories

::: callout-important
Make sure you're in the correct directory/using the correct path before using commands in this section.
:::

`rm`: This stands for "remove" and we can use it to remove files and directories.

To remove a directory and its contents, we would amend the code below to reflect the path and the directory we want and enter it into the terminal:

:::{.callout-warning}

## Using recursive mode 

Be cautious when using the rm -r command in Git Bash, as it recursively deletes the specified directory and all its contents. This action is irreversible, so double-check the directory path to avoid accidentally deleting important files or directories.
:::

```{bash }

> 

Reviewer: ```suggestion
```

rm -r path_to_the_directory/directory_name

> 

Reviewer: ```suggestion
```

```

> ### Removing files

Reviewer: I'd put this one first before the removing directories one I think.

Assuming we're already in the directory we want, we can remove a file from it by amending the code below to reflect the file name we want to remove, and its extension and then enter it into the terminal:

```{bash }

> 

Reviewer: ```suggestion
```

rm file_name.extention

> 

Reviewer: ```suggestion
```

```

You need to make sure to include the extension of the file in the command, or it will not work.

## Viewing File Contents

::: callout-important
Make sure you're in the correct directory/using the correct path before using commands in this section.
:::

The commands in this section allow you to display the contents of files, either in full or in parts, for easy reading and inspection.

To display the first 10 lines of a file, we would amend the code below to reflect the file name we want to view and its extension and then enter it into the terminal:

```{bash }

> 

Reviewer: ```suggestion
```

head file_name.extention

> 

Reviewer: ```suggestion
```

```

You can also customise the number of lines you want to view by adding `-number of lines` after `head`. For example, if we want to view the first 12 lines of a file, we use: 

```{bash }

> 

Reviewer: ```suggestion
```

head -12 file_name.extention

> 

Reviewer: ```suggestion
```

```

To display the bottom 10 lines of a file, we would amend the code below to reflect the file name we want to view and its extension and then enter it into the terminal:

```{bash }

> 

Reviewer: ```suggestion
```

tail file_name.extention

> 

Reviewer: ```suggestion
```

```

You can also customise the number of lines you want to view by adding `-number of lines` after `tail`. For example, if we want to view the last 3 lines of a file, we use: 

```{bash }

> 

Reviewer: ```suggestion
```

tail -3 file_name.extention

> 

Reviewer: ```suggestion
```

```

To display the contents of a file, we would amend the code below to reflect the file name we want to view and its extension and then enter it into the terminal:

```{bash }

> 

Reviewer: ```suggestion
```

cat file_name.extention

> 

Reviewer: ```suggestion
```

```

## Searching and Finding Files

### Finding files by patterns

::: callout-important
Make sure you're in the correct directory/using the correct path before using commands in this section.
:::

You can look for files in your repository that match a certain pattern by using `git grep`. We would amend the code below by replacing `"pattern"` with the term or pattern we want to search for in files.

```{bash }

> git grep "pattern"

Reviewer: Interesting, I didn't realise Git had their own implementation of grep. I think stick to the Bash one here and we can maybe have the Git version as an alternative in the Git glossary somewhere.
```suggestion
grep "pattern" filename
```
